### PR TITLE
Ebpf directory cache

### LINKED ISSUE #10854: Directory cache (eBPF plugin)
Operate systems keep a cache ([directory cache](https://www.usenix.org/legacy/publications/library/proceedings/usenix01/full_papers/kroeger/kroeger_html/node8.html)) on memory to speed up access to file. 

To help developers/SRE to improve the quality of the software and servers, Netdata will bring charts to monitor access to files, based in [IOvisor dcstat](https://github.com/iovisor/bcc/blob/master/tools/dcstat.py).

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
Fixes #10854

More details after https://github.com/netdata/kernel-collector/pull/209 to be merged.

##### Component Name
eBPF.plugin

### CODE REVIEW

**Review by vlvkobal — CHANGES_REQUESTED** [review]

**Comment by vlvkobal** [comment]
Please describe when `lookup_fast()` and `d_lookup()` are called and what is the meaning of reference, slow, and miss. It seems to me, that bpf counters, calculations, dimensions, and chart descriptions are messed up.

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+/**
+ * Update publish
+ *
+ * Update publish values before to write dimension.
+ *
+ * @param out   strcuture that will receive data.
+ * @param hit   calls to search a specific file
+ * @param miss  calls that did not find files
+ */
+void dcstat_update_publish(netdata_publish_dcstat_t *out, uint64_t hit, uint64_t miss)
```
[the remark this answers is not in the document]
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_dcstat.c]: I used the names from default algorithm, but I agree it can be better.

**Comment by thiagoftsm** [comment]
> Please describe when `lookup_fast()` and `d_lookup()` are called and what is the meaning of reference, slow, and miss. It seems to me, that bpf counters, calculations, dimensions, and chart descriptions are messed up.

I added a comment inside the source code to explain what the eBPF program is doing.  To change the dimension name is something that we will need to discuss with product team, because we are keeping the name to simplify the  users life when they compare our data with outputs from other software.

**Comment by vlvkobal** [comment]
There are no calls to the directory cache - that's quite strange. The other two charts (in the Applications section) are constantly growing.
image

**Comment by vlvkobal** [comment]
I think it's an incorrect placement in the menu
image

File: web/gui/dashboard_info.js [hunk 2]
```diff
@@ -1050,6 +1050,14 @@ netdataDashboard.context = {
         info: 'System calls for <a href="https://man7.org/linux/man-pages/man2/sync_file_range.2.html" target="_blank">sync_file_range()</a> permits fine control when synchronizing the open file referred to by the file descriptor fd with disk. This system call is extremely dangerous and should not be used in portable programs.'
     },
 
+    'filesystem.dc_hit_ratio': {
+        info: 'Percentage of file accesses that were present in the directory cache. 100% means that every file that was accessed was present in the directory cache. If files are not present in the directory cache 1)they are not present in the file system 2)the files were not accessed before. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```
Comment (vlvkobal) [inline web/gui/dashboard_info.js:1054]:
```suggestion
        info: 'Percentage of file accesses that were present in the directory cache. 100% means that every file that was accessed was present in the directory cache. If files are not present in the directory cache 1) they are not present in the file system, 2) the files were not accessed before. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```

File: web/gui/dashboard_info.js [hunk 3]
```diff
@@ -1243,6 +1251,22 @@ netdataDashboard.context = {
         info: 'Calls for function <code>udp_recvmsg</code>.'
     },
 
+    'apps.dc_hit_ratio': {
+        info: 'Percentage of file accesses that were present in the directory cache. 100% means that every file that was accessed was present in the directory cache. If files are not present in the directory cache 1)they are not present in the file system 2)the files were not accessed before. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```
Comment (vlvkobal) [inline web/gui/dashboard_info.js:1255]:
```suggestion
        info: 'Percentage of file accesses that were present in the directory cache. 100% means that every file that was accessed was present in the directory cache. If files are not present in the directory cache 1) they are not present in the file system, 2) the files were not accessed before. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```

**Comment by vlvkobal** [comment]
Chart titles and info strings for `apps.dc_reference` and `apps.dc_not_cache` contradict each other.

`Reference is when there is a file access` **!=** `Number of requests for files present inside directory cache`

`Slow is when there is a file access and the file is present in the filesystem but not in the directory cache` **!=** `Number of requests for files not present inside directory cache`

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+ *
+ * Update publish values before to write dimension.
+ *
+ * @param out           strcuture that will receive data.
+ * @param cache_access  number of access to directory cache.
+ * @param not_found     number of files not found on the file system
+ */
+void dcstat_update_publish(netdata_publish_dcstat_t *out, uint64_t cache_access, uint64_t not_found)
+{
+    calculated_number in_cache = (calculated_number) (((long long)cache_access) - ((long long)not_found));
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_dcstat.c:50]: `in cache` is `reference - slow`, not `reference - miss`.

**Comment by thiagoftsm** [comment]
> There are no calls to the directory cache - that's quite strange. The other two charts (in the Applications section) are constantly growing.
> image

Unfortunately I could not identify the charts you are showing, but I am assuming we are talking about these:

Screenshot_20210413_235154

On my  environment I am having something expected, after files to be moved for directory cache and I am not adding any new software to run, I am having few files outside `directory cache` and of course this implies they are not missed.

What was the kernel version used on your tests?

**Comment by thiagoftsm** [comment]
> I think it's an incorrect placement in the menu
> image

I was waiting for a product team decision, and I will assume that right now we will continue moving with `font awesome`.

**Comment by underhood** [comment]
> > I think it's an incorrect placement in the menu
> > image
> 
> I was waiting for a product team decision, and I will assume that right now we will continue moving with `font awesome`.

Not sure I understand what you try to say here. You want to keep it as new top-level item `filesystem` and then just update icon? If so please change it to `Filesystem` to match other items in menu

**Comment by thiagoftsm** [comment]
> Not sure I understand what you try to say here. You want to keep it as new top-level item `filesystem` and then just update icon? If so please change it to `Filesystem` to match other items in menu

You understood correctly, I already updated for `Filesystem` and I also selected an icon for it.

**Comment by vlvkobal** [comment]
> You understood correctly, I already updated for Filesystem and I also selected an icon for it.

Shouldn't it be just a family under the eBPF section in the menu?

**Comment by vlvkobal** [comment]
> On my environment I am having something expected, after files to be moved for directory cache and I am not adding any new software to run, I am having few files outside directory cache and of course this implies they are not missed.
> What was the kernel version used on your tests?

Well, another chart and `reference` is always zero

image

```
Linux arch-esxi 5.11.13-arch1-1 #1 SMP PREEMPT Sat, 10 Apr 2021 20:47:14 +0000 x86_64 GNU/Linux
/etc/lsb-release:LSB_VERSION=1.4
/etc/lsb-release:DISTRIB_ID=Arch
/etc/lsb-release:DISTRIB_RELEASE=rolling
/etc/lsb-release:DISTRIB_DESCRIPTION="Arch Linux"
/etc/os-release:NAME="Arch Linux"
/etc/os-release:PRETTY_NAME="Arch Linux"
/etc/os-release:ID=arch
/etc/os-release:BUILD_ID=rolling
/etc/os-release:ANSI_COLOR="38;2;23;147;209"
/etc/os-release:HOME_URL="https://www.archlinux.org/"
/etc/os-release:DOCUMENTATION_URL="https://wiki.archlinux.org/"
/etc/os-release:SUPPORT_URL="https://bbs.archlinux.org/"
/etc/os-release:BUG_REPORT_URL="https://bugs.archlinux.org/"
/etc/os-release:LOGO=archlinux
```
Is it even possible to enable/disable the directory cache?

**Comment by underhood** [comment]
I seem to miss the charts on ubuntu despite the thread running.... but let me double-check maybe I do sth. wrong/

**Review by underhood — APPROVED** [review]

**Comment by thiagoftsm** [comment]
> > You understood correctly, I already updated for Filesystem and I also selected an icon for it.
> 
> Shouldn't it be just a family under the eBPF menu section?

We are following a product request, their intention is move charts from eBPF to another sections.

**Comment by thiagoftsm** [comment]
> > On my environment I am having something expected, after files to be moved for directory cache and I am not adding any new software to run, I am having few files outside directory cache and of course this implies they are not missed.
> > What was the kernel version used on your tests?

I am upgrading a VM kernel now to test.

> Is it even possible to enable/disable the directory cache?

Yes, if we are talking about the plugin:

```
[ebpf programs]
    dcstat = no
```

now about the `directory cache`, this is not possible.

**Comment by vlvkobal** [comment]
> Yes, if we are talking about the plugin:

I was talking about the system directory cache, so the answer is

> now about the directory cache, this is not possible.

In that case `reference` should be >= `slow` >= `miss`. If `reference = 0`, then `slow` and `miss` should be zero. Thus, on my system `reference` is incorrect.

**Comment by thiagoftsm** [comment]
> > Yes, if we are talking about the plugin:
> 
> I was talking about the system directory cache, so the answer is
> 
> > now about the directory cache, this is not possible.
> 
> In that case `reference` should be >= `slow` >= `miss`. If `reference = 0`, then `slow` and `miss` should be zero. Thus, on my system `reference` is incorrect.

The last two commits addressed this issue, we are not having these problems now.

**Comment by vlvkobal** [comment]
After your three last commits, everything is constantly zero.

**Comment by thiagoftsm** [comment]
> After your three last commits, everything is constantly zero.

I tested now on `Slackware current` running the latest kernel version `5.11.16` and I have all charts running:

Screenshot_20210422_224602

I compiled the kernel from scratch using the following [config.gz](https://github.com/netdata/netdata/files/6361652/config.gz), please, can you give me your `/proc/config.gz` ? Or your `/boot/config`? Thanks!

**Comment by thiagoftsm** [comment]
@vlvkobal I retest the last commits on the following environments:

| Distribution | kernel version | Long term support |
|-----------------|--------------------|---------------------------|
|Slackware current | 5.11.16 | No |
|Slackware current | 5.10.32 | Yes|
|Slackware current | 5.4.114 | Yes|
|Debian 10.9 | 4.19.171-2 | Yes|
|CentOS Linux release 8.3.2011 | 4.18.0-240.22.1.el8_3.x86_64 | No|

With the exception of `Debian` that was a Virtual Machine, all the other environments were hosts with OS installed on them.

I could not identify any problem with charts using these distributions and kernels, when I receive your `config.gz`, I hope it will be possible to identify what is happening on your host.

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 5]
```diff
@@ -34,6 +34,12 @@ static ebpf_local_maps_t dcstat_maps[] = {{.name = "dcstat_pid", .internal_input
                                            .user_input = 0},
                                           {.name = NULL, .internal_input = 0, .user_input = 0}};
 
+static ebpf_specify_name_t dc_optional_name[] = { {.program_name = "netdata_lookup_fast",
+                                                  .function_to_attach = "lookup_fast",
+                                                   .optional = NULL,
+                                                   .retprobe = CONFIG_BOOLEAN_NO},
+                                                  {.program_name = NULL}};
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_dcstat.c:41]: Please format this properly.

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+    memset(dcstat_counter_publish_aggregated, 0, length*sizeof(netdata_publish_syscall_t));
+}
+
+/*****************************************************************
+ *
+ *  MAIN THREAD
+ *
+ *****************************************************************/
+
+
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_dcstat.c:551]:
```suggestion
```

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+ */
+void ebpf_dcstat_create_apps_charts(struct ebpf_module *em, void *ptr)
+{
+    UNUSED(em);
+    struct target *root = ptr;
+    ebpf_create_charts_on_apps(NETDATA_DC_HIT_CHART,
+                               "Percentage of files listed inside directory cache",
+                               EBPF_COMMON_DIMENSION_PERCENTAGE,
+                               NETDATA_APPS_DCSTAT_GROUP,
+                               NETDATA_EBPF_CHART_TYPE_STACKED,
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_dcstat.c:155]:
```suggestion
                               NETDATA_EBPF_CHART_TYPE_LINE,
```
Please do the same for `apps.cachestat_ratio` (in a separate PR?).

**Review by vlvkobal — CHANGES_REQUESTED** [review]

File: web/gui/dashboard_info.js [hunk 8]
```diff
@@ -1276,6 +1289,22 @@ netdataDashboard.context = {
...
+    'apps.dc_hit_ratio': {
+        info: 'Percentage of file accesses that were present in the directory cache. 100% means that every file that was accessed was present in the directory cache. If files are not present in the directory cache 1) they are not present in the file system, 2) the files were not accessed before. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
+    },
+
+    'apps.dc_reference': {
+        info: 'Counters of file accesses. <code>Reference</code> is when there is a file access. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
+    },
+
+    'apps.dc_not_cache': {
+        info: 'Counters of file accesses. <code>Slow</code> is when there is a file access and the file is present in the filesystem but not in the directory cache. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```
Comment (vlvkobal) [inline web/gui/dashboard_info.js:1301]:
```suggestion
        info: 'Counters of file accesses. <code>Slow</code> is when there is a file access and the file is not present in the directory cache. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```

File: web/gui/dashboard_info.js [hunk 9]
```diff
@@ -1079,6 +1084,14 @@ netdataDashboard.context = {
...
     },
 
+    'filesystem.dc_hit_ratio': {
+        info: 'Percentage of file accesses that were present in the directory cache. 100% means that every file that was accessed was present in the directory cache. If files are not present in the directory cache 1) they are not present in the file system, 2) the files were not accessed before. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
+    },
+
+    'filesystem.dc_reference': {
+        info: 'Counters of file accesses. <code>Reference</code> is when there is a file access and the file is present in the directory cache. <code>Miss</code> is when there is file access and the file is not found in the filesystem. <code>Slow</code> is when there is a file access and the file is present in the filesystem but not in the directory cache. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```
Comment (vlvkobal) [inline web/gui/dashboard_info.js:1092]:
```suggestion
        info: 'Counters of file accesses. <code>Reference</code> is when there is a file access and the file is present in the directory cache. <code>Miss</code> is when there is file access and the file is not found in the filesystem. <code>Slow</code> is when there is a file access and the file is not present in the directory cache. Read more about <a href="https://www.kernel.org/doc/htmldocs/filesystems/the_directory_cache.html" target="_blank">directory cache</a>.'
```

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 10]
```diff
@@ -0,0 +1,604 @@
...
+                               "Count file access.",
+                               EBPF_CACHESTAT_DIMENSION_HITS,
+                               NETDATA_APPS_DCSTAT_GROUP,
+                               NETDATA_EBPF_CHART_TYPE_STACKED,
+                               20101,
+                               ebpf_algorithms[NETDATA_EBPF_ABSOLUTE_IDX],
+                               root);
+
+    ebpf_create_charts_on_apps(NETDATA_DC_REQUEST_NOT_CACHE_CHART,
+                               "Access to files that were not present inside directory cache, but they are on filesystem.",
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_dcstat.c:170]:
```suggestion
                               "Access to files that were not present inside directory cache.",
```

**Review by vlvkobal — CHANGES_REQUESTED** [review]

File: collectors/ebpf.plugin/ebpf_dcstat.c (hunk 7 above)
Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_dcstat.c:155]: Yes, I will bring a fix for this case in a separated PR.

**Review by vlvkobal — DISMISSED** [review]

File: collectors/ebpf.plugin/ebpf_dcstat.c [hunk 11]
```diff
@@ -159,9 +159,9 @@ void ebpf_dcstat_create_apps_charts(struct ebpf_module *em, void *ptr)
 
     ebpf_create_charts_on_apps(NETDATA_DC_REFERENCE_CHART,
                                "Count file access.",
-                               EBPF_CACHESTAT_DIMENSION_HITS,
+                               EBPF_COMMON_DIMENSION_FILES,
                                NETDATA_APPS_DCSTAT_GROUP,
-                               NETDATA_EBPF_CHART_TYPE_STACKED,
+                               NETDATA_EBPF_CHART_TYPE_LINE,
```
Comment (vlvkobal) [inline collectors/ebpf.plugin/ebpf_dcstat.c:164]: Why did you changed that?

Comment (vlvkobal) [reply collectors/ebpf.plugin/ebpf_dcstat.c:164]: `STACKED` was correct.

Comment (thiagoftsm) [reply collectors/ebpf.plugin/ebpf_dcstat.c]: Sorry my fault, I changed in the wrong branch a test I was doing. :/

**Review by vlvkobal — APPROVED** [review]

**Comment by ilyam8** [comment]
About adding `/s` prefix for [incremetal](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#dimension) dimensions.

I think this is important and not adding it is misleading. Let me give you an example when `update_every` is 5.

| Time        | Collected value | Stored value (interpolation) | Meaning                |
| :---        |    :----:       |    :---:                     | :---:                  | 
| t0 (0)      | 10              | -                            | -                      |
| t1 (5)      | 20              | 2                            | Per second rate t0->t1 |
| t2 (10)     | 30              | 2                            | Per second rate t1->t2 |
| t3 (15)     | 40              | 2                            | Per second rate t2->t3 |

As you see the stored value doesn't mean that it happened 2 events, but the per second rate was 2 events per second.

**Comment by thiagoftsm** [comment]
> About adding `/s` suffix for [incremetal](https://learn.netdata.cloud/docs/agent/collectors/plugins.d#dimension) dimensions.
> 
> I think this is important and not adding it is misleading. Let me give you an example when `update_every` is 5.
> Time 	Collected value 	Stored value (interpolation) 	Meaning
> t0 (0) 	10 	- 	-
> t1 (5) 	20 	2 	Per second rate t0->t1
> t2 (10) 	30 	2 	Per second rate t1->t2
> t3 (15) 	40 	2 	Per second rate t2->t3
> 
> As you see the stored value doesn't mean that it happened 2 events, but the per second rate was 2 events per second.

I agree with you, but this PR does not have `incremental` dimensions, they are treated as absolute, because we do some math to avoid big gaps on charts like other companies deliver.

**Review by underhood — APPROVED** [review]

**Pull request merged**